Amesos2: Tacho Wrapper Setup

MicheldeMessieres opened the pull request:
This is in progress.
The current status is that the wrapper should run and pass for real/complex.

Some open issues to resolve:

The current Tacho complex unit test matrix was setup with all imaginary 0.
The wrapper is currently using the same.

I left a transpose option in the test but it currently does nothing - I will delete it if we don’t expect to test that.

I’m not sure how to best setup the idx (permutation) parameter - currently just assigned 0,1,2… as in the examples. I’m storing this as a persistent value but there might be a more elegant way to handle this if it will be a default state.

I added the OpenMP DeviceSpaceType when OpenMP is enabled.
I’ve left the Amesos2 test as MPI processes 4 but this runs very slowly on OpenMP (about 80s).
I noted changing the demo test to MPI 4 is also somewhat slow (3s).
Wasn’t sure if we should go with MPI 1 and did not investigate this yet.
Some guidance on how/if we want thread options setup will help, do we want a parameter, etc.

Other issues are probably related to how the Kokkos::Views are set up to pass to Tacho. I think this is best discussed by inspecting:
 Amesos2_Tacho_decl.hpp
 Amesos2_Tacho_def.hpp

I modeled some 1 proc optimizations on Basker but there are some differences which may need updating as well.

kyungjoo-kim commented:
1. For the complex version test, I separately test for more realistic test cases but I cannot commit the matrices until we have separate test repo is provided. 

2. This is Cholesky factorization (symmetric positive definite). So, the transpose of the matrix is essentially same for double precision. But we need to apply conjugate when transpose is applied to a complex matrix. Do we have the transpose option before we factorize matrix ? or in the solve phase. Since transpose(A^-1) = transpose(A)^-1 when A is hermitian, there should be a way to apply transpose solve.

3. I don't understand about the permutation parameter. Is it only related to amesos2 ?  

4. Do you mean the oversubscription of threads when you run mpi + opnmp together ?

MicheldeMessieres commented:
1. I'll update the complex matrix for the wrapper once we have decided what we want used, though I wasn't sure what you meant about needing a separate test repo. I saw both a real and complex matrix in the Tacho test folder but wasn't sure why the complex was all 0 imaginary - I think that is meant as a placeholder? I'm currently using the same forms for Amesos2.

2. We probably don't need any transpose options for the Amesos2 wrapper if Tacho doesn't have any transpose options to test. If we did need to do anything with the wrapper we can apply at factorization or the solve phase as needed.

3. I mean the input permutation (perm, peri) for SymbolicTools and NumericTools. Currently I am just generating the simple case (0,1,2...) which may be fine for the purposes of the wrapper.

4. Since we're testing Tacho right now with MPI 1 it probably makes sense for me to change the Amesos2 wrapper to match. I wasn't sure if we wanted to have any tests of the Amesos2 wrapper set for more than 1 mpi.

kyungjoo-kim commented:
1. separate data repo has been discussed here. as trilinos do not encourage to add big test file in the repo (however, they still want to test the code with a more realistic test problem in cdash), they might create another repo for this purpose collecting test problems and big data files in separate repo for automatic test.

2. if not necessary, let's not create the transpose solve interface. 

3. the permutation array should comes from graph reordering library i.e., metis. the symbolic phase should include graph reordering part. If you use the solver interface in Tacho_Solver.hpp, it has only analyze, factorize, and solve. the analyze phase includes reordering by metis. right now your interface uses somewhat low-level part of my code.

MicheldeMessieres commented:
@kyungjoo-kim I've updated the PR to use the API from Tacho_Solver so I think this will be closer to what you expected.

I marked a few places as TODO which could be helpful to discuss but the current form is working for complex on/off.  I've removed the transpose option. I added some commented out code to begin setting up parameters but wanted to confirm which we want, if any. I've got the type maps setup now to use Kokkos::complex, not std::complex. I think that is the only way I can currently do this using the Tacho_Solver code but I'm not certain if this is the best way to setup the wrapper. This may be fine though.

Note that the Tacho code currently generates a lot of shadow warnings (if compiling with flag Shadow).

Also added a 2nd commit with proposed edits to the Tacho tests so they can be turned on more easily. That caused me some difficulties because Tacho tests may build fine if something else like Amesos2 is on (since it pulls in the dependencies) but then break when turning that off.

kyungjoo-kim commented:
I thought that @srajama1 handle this pull request but I probably need to do this. 

This seems to be mostly fine and I need to checkout and compile the pull request and see how it works. 

One thing I would like to mention is that Tacho uses predefined ordinal and size type compilated by cmake. I wonder how the amesos interface and its ETI consider these types. I will come back after I compile the code. 

```
#if defined( TACHO_USE_INT_INT )                                                                              
    typedef int    ordinal_type;                                                                              
    typedef int    size_type;                                                                                 
#elif defined( TACHO_USE_INT_SIZE_T )                                                                         
    typedef int    ordinal_type;                                                                              
    typedef size_t size_type;                                                                                 
#else                                                                                                         
    typedef int    ordinal_type;                                                                              
    typedef size_t size_type;                                                                                 
#endif     
```

kyungjoo-kim commented:
If I want to test the pull request, what is the additional cmake option that I need to add ?

MicheldeMessieres commented:
@kyungjoo-kim  hi - I think just the following::

-DTrilinos_ENABLE_Amesos2:BOOL=ON \
-DAmesos2_ENABLE_Tacho:BOOL=ON \
-DAmesos2_ENABLE_TESTS:BOOL=ON \
-DTrilinos_ENABLE_ShyLUTacho:BOOL=ON \

kyungjoo-kim commented:
Seems that it causes a trouble of different size_type and ordinal_type. The ordinal and size types are controlled by -DTacho_ENABLE_INT_INT:BOOL=ON. This uses ordinal_type(int) and size_type(int). If the flag is off, then ordinal_type(int) and size_type(size_t). This needs to be fixed.

```
/home/kyukim/Work/lib/trilinos/tech-x/packages/kokkos/core/src/Kokkos_View.hpp: In instantiation of ‘Kokkos::View<DataType, Properties>& Kokkos::View<DataType, Properties>::operator=(const Kokkos::View<RT, RP ...>&) [with RT = int*; RP = {Kokkos::Serial}; DataType = long int*; Properties = {Kokkos::Serial}]’:
/home/kyukim/Work/lib/trilinos/tech-x/packages/shylu/tacho/core/src/TachoExp_GraphTools_Metis.hpp:49:15:   required from here
/home/kyukim/Work/lib/trilinos/tech-x/packages/kokkos/core/src/Kokkos_View.hpp:1098:7: error: static assertion failed: Incompatible View copy assignment
       static_assert( Mapping::is_assignable , "Incompatible View copy assignment" );
       ^
/home/kyukim/Work/lib/trilinos/tech-x/packages/kokkos/core/src/Kokkos_View.hpp: In instantiation of ‘Kokkos::View<DataType, Properties>::View(const Kokkos::View<RT, RP ...>&) [with RT = long int*; RP = {Kokkos::Serial}; DataType = int*; Properties = {Kokkos::Serial}]’:
/home/kyukim/Work/lib/trilinos/tech-x/packages/shylu/tacho/core/unit-test/Tacho_TestSymbolic.hpp:83:25:   required from here
/home/kyukim/Work/lib/trilinos/tech-x/packages/kokkos/core/src/Kokkos_View.hpp:1088:7: error: static assertion failed: Incompatible View copy construction
       static_assert( Mapping::is_assignable , "Incompatible View copy construction" );
       ^
/home/kyukim/Work/lib/trilinos/tech-x/packages/kokkos/core/src/Kokkos_View.hpp: In instantiation of ‘Kokkos::View<DataType, Properties>& Kokkos::View<DataType, Properties>::operator=(const Kokkos::View<RT, RP ...>&) [with RT = long int*; RP = {Kokkos::Serial}; DataType = int*; Properties = {Kokkos::Serial}]’:
/home/kyukim/Work/lib/trilinos/tech-x/packages/shylu/tacho/core/src/TachoExp_SymbolicTools.hpp:516:15:   required from ‘Tacho::Experimental::SymbolicTools::SymbolicTools(CrsMatBaseType&, GraphToolType&) [with CrsMatBaseType = Tacho::Experimental::CrsMatrixBase<Kokkos::complex<double>, Kokkos::Serial>; GraphToolType = Tacho::Experimental::GraphTools_Metis]’
/home/kyukim/Work/lib/trilinos/tech-x/packages/shylu/tacho/core/unit-test/Tacho_TestSymbolic.hpp:171:25:   required from here
/home/kyukim/Work/lib/trilinos/tech-x/packages/kokkos/core/src/Kokkos_View.hpp:1098:7: error: static assertion failed: Incompatible View copy assignment
       static_assert( Mapping::is_assignable , "Incompatible View copy assignment" );
       ^
```

MicheldeMessieres commented:
@kyungjoo-kim I rebased to develop and force pushed so this is synched now to the latest.

However I can't replicate the error you mentioned. I tried with -DTacho_ENABLE_INT_INT on and off and I verified that I was seeing the change in the sizes of the types used by the Amesos2 wrapper. It appears to be working for both cases. I had setup the Amesos2 wrapper to read the types from TachoExp_Util.hpp via Tacho.hpp so I think it should be bringing in the proper types based on defines from ShyLUTacho_config.h.

Can you give me your full config for generating the error? I will try to replicate from that. Thanks.

kyungjoo-kim commented:
Hi I still see the error and I use the following branch and revision

I use ``checkin-test-sems.sh``. Create a separate build directory and create a symbolic link located in ``Trilinos/cmake/std/sems/checkin-test-sems.sh``. Run the script

```
./checkin-test-sems.sh --configure --allow-no-pull --enable-packages='ShyLUTacho;
```
This generates cmake files.
```
./do-configure.base \                                                                                                                                                                  
-DTrilinos_ENABLE_Fortran:BOOL=OFF \                                                                                                                                                   
-DTrilinos_ENABLE_Amesos2:BOOL=ON \                                                                                                                                                    
-DAmesos2_ENABLE_Tacho:BOOL=ON \                                                                                                                                                       
-DAmesos2_ENABLE_TESTS:BOOL=ON \                                                                                                                                                       
-DTrilinos_ENABLE_ShyLUTacho:BOOL=ON \                                                                                                                                                 
-DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=OFF \                                                                                                                                     
-DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=OFF \                                                                                                                                  
-DTrilinos_ENABLE_PyTrilinos:BOOL=OFF \                                                                                                                                                
-DTrilinos_ENABLE_Claps:BOOL=OFF \                                                                                                                                                     
-DTrilinos_ENABLE_TriKota:BOOL=OFF \                                                                                                                                                   
"$@"  
```
```
cmake \                                                                                                                                                                                
-DTrilinos_TRIBITS_DIR:PATH=/home/kyukim/Work/lib/trilinos/tech-x/cmake/tribits \                                                                                                      
-DTrilinos_ENABLE_TESTS:BOOL=ON \                                                                                                                                                      
-DTrilinos_TEST_CATEGORIES:STRING=BASIC \                                                                                                                                              
-DTrilinos_ALLOW_NO_PACKAGES:BOOL=OFF \                                                                                                                                                
-DDART_TESTING_TIMEOUT:STRING=300.0 \                                                                                                                                                  
-DBUILD_SHARED_LIBS=ON \                                                                                                                                                               
-DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON \                                                                                                                                   
-DTrilinos_ENABLE_SECONDARY_TESTED_CODE:BOOL=OFF \                                                                                                                                     
-DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/MpiReleaseDebugSharedPtSettings.cmake,cmake/std/BasicCiTestingSettings.cmake \                                                      
"$@" \                                                                                                                                                                                 
/home/kyukim/Work/lib/trilinos/tech-x/cmake/tribits/ci_support/../../..                                                                                                                
                                                                                                                                                                                       
                                                                                                                                                                                       
                                                                                                                                                                                       
                                                                                                                                                                                       
                                                                                                                                                                                       
                                                                                                                                                                                       
                                                                                                                                                                                       
                                      
```



```
[kyukim @bread] tech-x > git branch -a 
* amesos2Tacho
  master

[kyukim @bread] tech-x > git log 
      1 commit cf9498575635aa831e829aa583458556f4866fdc
      2 Merge: 32cf189 2764757
      3 Author: Kyungjoo Kim <kyukim@sandia.gov>
      4 Date:   Sun Sep 10 13:26:49 2017 -0600
      5 
      6     Merge branch 'amesos2Tacho' of https://github.com/Tech-xCorp/Trilinos into amesos2Tacho
      7 
      8 commit 276475773c7cac695ec9b6096b4a1b90a7d1d287
      9 Author: micheldemessieres <michel@demessieres.com>
     10 Date:   Thu Aug 24 11:50:33 2017 -0400
     11 
     12     Tacho: Update test cmake dependencies
     13     
     14     Propose some updates to TEST_REQUIRED_DEP_TPLS for shylu/tacho.
     15       Added BLAS, LAPACK
     16     
     17     Propose some updated to TEST_REQUIRED_DEP_PACKAGES for shylu/tacho
     18       Moved TrilinosSS and Gtest from optional to required
     19     
     20     Without above changes build errors may or may not occur depending on
     21     which other packages are enabled.
```

MicheldeMessieres commented:
@kyungjoo-kim I am trying SEMS on a sandia machine. If I am on the current develop branch (not amesos2Tacho) I can replicate some of the errors you mentioned by just doing the following to build the Tacho tests/examples:

./checkin-test-sems.sh -j10 --local-do-all --enable-packages='ShyLUTacho' --extra-builds=MPI_RELEASE_DEBUG_SHARED_PT --default-builds='' --no-enable-fwd-packages

However I can't build Tacho using the command you described - I would expect to need something like the above since Tacho is an EX package. Is Amesos2 getting turned on from the SEMS command or are you turning that on somewhere else? I think I need more info to replicate what you are doing with SEMS.

However I think i may see the same errors for SEMS with the develop branch, just building ShyLUTacho, so perhaps this is not an amesos2Tacho branch issue? Can you confirm if you can build Tacho tests/examples on SEMS using the current develop branch and if there is anything special I need to do to see that? I'll need to figure out how to get Tacho building for SEMS before I proceed to test the amesos2Branch for SEMS. Local gcc and clang builds I've tried seem fine. Thanks.

kyungjoo-kim commented:
Oh.. you should also add ``--enable-packages='ShyLUTacho,Amesos2'``. Check the configuration files generated from the check in script if it includes the cmake option that you added for compiling with Amesos2. BTW, I got this error from your branch (not from develop).

MicheldeMessieres commented:
@kyungjoo-kim Can you confirm if develop builds Tacho correctly for SEMS for tests/examples? I could not build Tacho tests/examples for develop branch using SEMS but saw the same errors so the issue seems to be unrelated to this new work. Or I don't know how to configure the develop branch. I need to resolve the develop branch issues first.

kyungjoo-kim commented:
Oh... I saw that the check-in-script disable Tacho as it is labed with "EX". All ShyLU subpackages are EX stage now. 

Using the current checkin script, we need to enable Tacho manually. Please do as follows
```
./checkin-test-sems.sh --enable-packages="Amesos2" --configure --allow-no-pull 
```
This will generate do-configure and do-configure.base in MPI_RELEASE_DEBUG_SHARED_PT directory. Please modify the configure file to include 
```
-DTrilinos_ENABLE_ShyLUTacho:BOOL=ON \                                                                                
-DShyLUTacho_ENABLE_EXAMPLES:BOOL=ON \                                                                                
-DShyLUTacho_ENABLE_TESTS:BOOL=ON \                                                                                   
-DTacho_ENABLE_INT_INT:BOOL=ON \ 
```

Then, 
```
rm CMakeCache.txt; rm -rf CMakeFiles;
source your-trilinos-directory/cmake/load_ci_sems_dev_env.sh
do-configure
```

This will generate the cmake script identical to mine. 


BTW, while reproducing the compile script, I found that the problem is on my side. The issue is from metis ``idx_t`` type is sometimes 64 bit integer and sometimes 32 bit integer according to the linking library. 

I will get back to you in an hour.

kyungjoo-kim commented:
Hi, 

I pushed the fix on develop branch. I tested your branch with my modification on Tacho. Please sync with develop branch and check if it is still okay if you locally made additional changes.

kyungjoo-kim commented on packages/amesos2/src/Amesos2_Tacho_def.hpp:
```diff
@@ -0,0 +1,371 @@
+      row_ptr = size_type_array(this->rowptr_.getRawPtr(), this->globalNumRows_ + 1);
+      cols = ordinal_type_array(this->colind_.getRawPtr(), this->globalNumNonZeros_);
+    }
+
```
row_ptr and cols are type defined from Tacho. Tacho does not have template parameters for local/global ordinals. 

How do I know that this->rowptr_ has the same type while Tpetra can instanciate different combinations of local/global ordinals ?

MicheldeMessieres replied:
rowptr_ is setup to be type Tacho::Experimental::size_type as a Teuchos::Array and this is filled by Util::get_crs_helper in the loadA_impl so I think this handles all the type needs. I had some optimizations for a single process case (similar to Basker) which Siva and I discussed briefly but I wasn't sure if this can benefit Tacho right now unless we changes Tacho to handle different types coming in. Something I will need to get further clarification on. I have temporarily left that commented out for discussion.

kyungjoo-kim commented on packages/amesos2/src/Amesos2_Tacho_def.hpp:
```diff
@@ -0,0 +1,371 @@
+        solve_array_t b(&bValues.getRawPtr()[j*i_ld_rhs], this->globalNumRows_, n);
+        solve_array_t t("t", this->globalNumRows_, n);
+        data_.solver.solve(x, b, t);
+
```
Consider that this solve routine is called many times. x and b are just wrapping pointers and it is okay. Constructing the temporary workspace of "t" is not necessary to be called multiple times if it is created before. Can you put this as member variable and construct the array when it is needed ?

mhoemmen replied:
`rowptr`'s entry type is actually a function of the Kokkos execution space's default `size_type`.

kyungjoo-kim replied:
@mhoemmen 

I think that this comment is about my previous comment. 

In my code, I do not inherit oridnal_type and size_type from Kokkos execution space but they are simply "int" and "size_t" unless cmake configuration enforces to use int as its size_type (SIERRA uses int only).

@MicheldeMessieres 

Do we have full unit tests that cover all ordinal type combinations ?

MicheldeMessieres replied:
Currently Amesos2 is not testing both Tacho_ENABLE_INT_INT ON/OFF. In the current setup this could be tested by manually setting ON/OFF for SEMS. If we wanted to have Amesos2 run two different tests for both cases we might want to rethink how this is setup since it is currently controlled at the Tacho level and coded into ShyLUTacho_config.h. I'm not sure what the right design is to avoid confusion what the config options mean and where we would want Amesos2 controlling this.

MicheldeMessieres replied:
I've pushed some new changes which move the workspace t to be allocated just once. Also some other changes to remove some typedefs that were not needed. This was checked on SEMS to the latest develop for complex on/off.

kyungjoo-kim replied:
It is okay. Tacho_ENABLE_INT_INT is for seirra and they do not use tacho through Amesos2. By default, we test Tacho_ENABLE_INT_INT=OFF. if this is not configured in cmake, tacho size_type is always size_t.

kyungjoo-kim commented:
@srajama1 

As all subpackages under Shylu are "EX" state and it is not compiled when one uses check-in script. 

Do we need to change the label so that those packages are included in check-in script ?

MicheldeMessieres commented:
@kyungjoo-kim Tried the new code on SEMS and it passes fine now for Tacho and Amesos2. Will reply to your comments above and make the necessary changes.

kyungjoo-kim commented on packages/amesos2/src/Amesos2_Tacho_def.hpp:
```diff
@@ -208,21 +208,11 @@ TachoSolver<Matrix,Vector>::solve_impl(const Teuchos::Ptr<MultiVecAdapter<Vector
-      data_.solver.solve(x, b, t);
+
+      data_.solver.solve(x, b, workspace_);
 
```
So, Do we always assume that we have a single rhs ? Is Tpetra disable to use multiple rhs ? If nrhs is not 1, I would do 
```
const int nrhs = from_tpetra_multivector_num_columns;
if (t.dimension_0() < this->globalNumRows() || t.dimension_1() < nrhs) {
  t = solve_array_t("t", this->globalNumRows, nrhs);
}
data_.solver.solve(x, b, t);
```

MicheldeMessieres replied:
I will fix this so the solver is not passing them through one at a time. I'll build the workspace as you suggested so it only resizes when necessary.

MicheldeMessieres replied:
@kyungjoo-kim Fixed so it passes the full set of vectors to Tacho now.

kyungjoo-kim commented on packages/amesos2/src/Amesos2_Tacho_def.hpp:
```diff
@@ -109,28 +112,20 @@ TachoSolver<Matrix,Vector>::symbolicFactorization_impl()
     }
     else
+*/
     {
```
Overall, we can do both. 
```
if (std::is_same<size_type,typename decltyp(sp_rowptr)::non_const_value_type>::value) {
  // use view directly via shallow copy
} else {
  // convert
}
```

MicheldeMessieres replied:
If we use the direct shallow copy we still need that code to compile for the non optimized cases. I have updated with a working method but have to use a force cast to achieve this. I'm not sure if this is the best way to handle this situation. I've tested the current in SEMS and seems ok for complex on and off.

kyungjoo-kim commented:
@MicheldeMessieres  

I am fine with the modification. One last question. Do we need to use Teuchos Array in the Tpetra interface ?

MicheldeMessieres commented:
@kyungjoo-kim Thanks. I think we just need the Teuchos arrays for Nprocs > 1. They are loaded when we call do_get  on the MatrixAdapter. For 1 proc, the optimization turns on and the matrix is read directly. Siva and I briefly discussed having this follow the format of Basker but I think some further discussion will be useful.

kyungjoo-kim commented:
Merged.
